File: client/mobile/lib/services/auth_service.dart
import '../models/user_model.dart';
import 'api_client.dart';
import 'storage_service.dart';

class AuthResult {
  final UserModel user;
  final String token;

  const AuthResult({required this.user, required this.token});
}

class AuthService {
  final ApiClient _client = ApiClient();

  Future<AuthResult> register({
    required String email,
    required String password,
    required String fullName,
    required String username,
  }) async {
    try {
      final response = await _client.dio.post(
        '/auth/register',
        data: {
          'email': email.trim(),
          'password': password,
          'full_name': fullName.trim(),
          'username': username.trim(),
        },
      );

      final data = response.data['data'] as Map<String, dynamic>;
      final user = UserModel.fromJson(data['user'] as Map<String, dynamic>);
      final token = data['token'] as String;

      await StorageService.saveToken(token);
      await StorageService.saveUser(user);

      return AuthResult(user: user, token: token);
    } catch (e) {
      throw _client.extractErrorMessage(e);
    }
  }

  Future<AuthResult> login({
    required String email,
    required String password,
  }) async {
    try {
      final response = await _client.dio.post(
        '/auth/login',
        data: {
          'email': email.trim(),
          'password': password,
        },
      );

      final data = response.data['data'] as Map<String, dynamic>;
      final user = UserModel.fromJson(data['user'] as Map<String, dynamic>);
      final token = data['token'] as String;

      await StorageService.saveToken(token);
      await StorageService.saveUser(user);

      return AuthResult(user: user, token: token);
    } catch (e) {
      throw _client.extractErrorMessage(e);
    }
  }

  Future<UserModel> getMe() async {
    try {
      final response = await _client.dio.get('/auth/me');
      final data = response.data['data'] as Map<String, dynamic>;
      final user = UserModel.fromJson(data);
      await StorageService.saveUser(user);
      return user;
    } catch (e) {
      throw _client.extractErrorMessage(e);
    }
  }

  Future<void> logout() async {
    try {
      await _client.dio.post('/auth/logout');
    } catch (_) {
      // Ignore network failure on logout
    } finally {
      await StorageService.clear();
    }
  }
}

File: client/web/src/services/authService.js
import { api, extractErrorMessage } from './api';
import { APP_CONFIG } from '../config/constants';

export const authService = {
  /**
   * Log in user with email & password
   */
  async login({ email, password }) {
    try {
      const response = await api.post('/auth/login', {
        email: email.trim(),
        password,
      });

      const { user, token } = response.data.data;
      if (token) {
        localStorage.setItem(APP_CONFIG.storageKeys.token, token);
        localStorage.setItem(APP_CONFIG.storageKeys.user, JSON.stringify(user));
      }
      return { user, token };
    } catch (error) {
      throw new Error(extractErrorMessage(error));
    }
  },

  /**
   * Register new user (Client or Provider)
   */
  async register({
    email,
    password,
    fullName,
    username,
    phone,
    locationCity,
    role = 'client',
    headline,
  }) {
    try {
      const response = await api.post('/auth/register', {
        email: email.trim(),
        password,
        full_name: fullName.trim(),
        username: username.trim(),
        phone: phone ? phone.trim() : undefined,
        location_city: locationCity ? locationCity.trim() : 'Addis Ababa',
        role,
        headline: headline ? headline.trim() : undefined,
      });

      const { user, token } = response.data.data;
      if (token) {
        localStorage.setItem(APP_CONFIG.storageKeys.token, token);
        localStorage.setItem(APP_CONFIG.storageKeys.user, JSON.stringify(user));
      }
      return { user, token };
    } catch (error) {
      throw new Error(extractErrorMessage(error));
    }
  },

  /**
   * Fetch current authenticated user
   */
  async getMe() {
    try {
      const response = await api.get('/auth/me');
      const user = response.data.data;
      localStorage.setItem(APP_CONFIG.storageKeys.user, JSON.stringify(user));
      return user;
    } catch (error) {
      throw new Error(extractErrorMessage(error));
    }
  },

  /**
   * Sign out
   */
  async logout() {
    try {
      await api.post('/auth/logout');
    } catch (_) {
      // Ignore network errors on logout
    } finally {
      localStorage.removeItem(APP_CONFIG.storageKeys.token);
      localStorage.removeItem(APP_CONFIG.storageKeys.user);
    }
  },
};
